### ISSUE TITLE
Memory leaks when developer tools are open in Chrome

### ISSUE THREAD

**Issue by brgrz** [angular-ui/angular-google-maps#1524]
I finally nailed this down. AGM's memory consumption skyrockets when Chrome dev tools are opened. It'd go from 200 MB to 600 MB instantly and then after a few pans/zooms it would go to 800-1GB. 

User experience quickly becomes a pain, UI starts freezing/hanging.

I have tested this with 2.1.6 and 2.2.1 with the Example project from source. The same happens when using the demo online.

What could this relate to? Some debug settings in AGM code?

**Comment by nmccready** [comment]
I am not seeing this; I do see some memory not being collected but nothing as extreme as you state. I need more steps to reproduce. Help me help you.

**Comment by nmccready** [comment]
I did several zoom in/outs and pans and generated 1000 clusters and non cluster markers. Memory never got over 430 MB. It did collect later on back down to 406MB.

**Comment by brgrz** [comment]
I too can't reproduce it right now when I'm in a different user profile on my dev machine but I did notice it a kit before. Strange. Will continue to monitor the behavior.

**Comment by spyridonas** [comment]
Since you reference my issue I can only provide the following info : my memory goes through the roof because when I update the markers the old ones don't get destroyed. So my scope has 3 event handlers, then 6, then 9, 12 etc. Imagine that increase every half second and you got problem. Can you create a clearAll(); ? Thanks.

**Comment by nmccready** [comment]
There is already a Plurals.clear via control. Which MarkersParent inherits. Also you guys really havent stated exactly which directive is your problem. I assume it is ui-gmap-markers but it could be ui-gmap-marker,

https://github.com/angular-ui/angular-google-maps/blob/master/src/coffee/directives/api/plural.coffee#L14-L15

**Comment by nmccready** [comment]
It does seem AngularJS Batarang is more stable then it has been in a while. I do see that some scopes are not cleaning up.

**Comment by nmccready** [comment]
To me looking at the scopes that are around it appears to be windows, or window lingering.

**Comment by spyridonas** [comment]
Thanks, i do use ui-gmap-markers. I don't know if brgrz has the same problem , maybe he can try and implement the Plurals.clear ?

**Comment by nmccready** [comment]
To me markers does not appear to be a problem but if your using windows with it then you may have an issue.

**Comment by nmccready** [comment]
I am coming to these assumptions by looking at example.js/.html , all the leaky scopes have either `show, isIconVisibleOnClick, or template` which are all window stuff.

**Comment by nmccready** [comment]
I found and fixed one of the leaks.

**Comment by nmccready** [comment]
Commit above has fixes; I will PR it later.

**Comment by spyridonas** [comment]
Thank you for taking the time to look at the issue, using your commit i did use o noticeable improvement. There are still more leaks though.

**Comment by spyridonas** [comment]
After trying to find the issue a possible finding is that the window html cache doesn't get removed. Some answers from stackoverflow "The entries in the cache object do only get removed if jQuery knows that the elements have been removed. That is the elements have to be removed with jQuery." and "I changed the plugin code to unbind the event handlers and remove the data in the plugin's destroy() method, that eventually removed the memory leak."

**Comment by nmccready** [comment]
As for the cache issues I recommend replacing angulars cache with this https://github.com/jmdobry/angular-cache , it automatically does an LRU cache based on your config.

**Comment by nmccready** [comment]
Closing w/PR #1528

**Comment by spyridonas** [comment]
When using the angular-cache you stated above the html cache of infoWindow still remains between refreshes ( I don't do any caching, i mean the cache you have inside your plugin)

**Comment by nmccready** [comment]
Ok, mis interpreted what you said. So you mean the templated window is stuck in scope. Yeah I did see that and I tried to fix it but got nowhere. So I don't have a fix for that at the moment. Make a new ticket for that. Be nice if someone else dived into this.

**Comment by nmccready** [comment]
@spyridonas I'll reopen this but you need to be clear on what code lines help you. Just rambling will get you no where. Please learn git conventions via markdown for copying code lines and using code blocks.

[https://github.com/adam-p/markdown-here/wiki/Markdown-Cheatsheet](markdown cheatsheat)

**Comment by nmccready** [comment]
Also what is leaking / remaining besides the template stuff. 
Just stating:

>  Can somebody implement this fix properly and pr it ?

Is a sure way to get no one to help you. Because basically it means this:

![13bbe1a8d29dbe9d82ff769f6561c66a285a64ed30630445c138a664983a91d4](https://cloud.githubusercontent.com/assets/583593/10073235/689080e8-6295-11e5-8007-8b7a9aca3074.jpg)

**Comment by spyridonas** [comment]
Ok i will try my best, here are my modifications

src\coffee\directives\api\models\child\window-child-model.coffee , line 200 (method destroy)
```
destroy: (manualOverride = false) =>
          @remove()
          if (@scope? and not @scope?.$$destroyed) and (@needToManualDestroy or manualOverride)
            @scope.$destroy()
          @hideWindow()
          @removeEvents @listeners
          @listeners.length = 0
          delete @gObject
          delete @opts
```

src\coffee\directives\api\managers\marker-manager.coffee, line 56 (method destroy)
```
destroy: (manualOverride = false) =>
          @remove()
          if (@scope? and not @scope?.$$destroyed) and (@needToManualDestroy or manualOverride)
            @scope.$destroy()
          @hideWindow()
          @removeEvents @listeners
          @listeners.length = 0
          delete @gObject
          delete @opts
          @clear
```
src\coffee\directives\api\managers\marker-manager.coffee, line 59 (method clear)

```
    clear: =>
      @gMarkers.each (gMarker) ->
        gMarker.setMap null
      delete @gMarkers
      @hideWindow()
      @removeEvents @listeners
      @listeners.length = 0
      delete @gObject
      delete @gMarkers
      delete @opts
      @gMarkers = new PropMap()
    
```

**Comment by nmccready** [comment]
You know github has line highlighting built into the urls if you click a line.

https://github.com/angular-ui/angular-google-maps/blob/master/src/coffee/directives/api/models/child/window-child-model.coffee#L200

Notice `#L200` .

Also when you go to the root page of any github project if you hit the keyboard 'T' key it will open up a fast search for files in a project. Example hit T then type window-ch and window-child-model is right there.

**Comment by nmccready** [comment]
A child should not be touching the manager from the clean perspective. But I thought you stated that you made progress copying and pasting something.. what did you try? Whats  lines of js did u move?

**Comment by spyridonas** [comment]
I copy-pasted all the lines from https://github.com/angular-ui/angular-google-maps/blob/master/src/coffee/directives/api/models/child/window-child-model.coffee#L193 and i think all of my leaks are fixed.

**Comment by nmccready** [comment]
yeah but where did you paste it too? Thinking the leak is fixed is not enough. More then likely many times you break something somewhere else.

**Comment by spyridonas** [comment]
I pasted the lines on 

https://github.com/angular-ui/angular-google-maps/blob/master/src/coffee/directives/api/models/child/window-child-model.coffee#L200

https://github.com/angular-ui/angular-google-maps/blob/master/src/coffee/directives/api/managers/marker-manager.coffee#L56 

https://github.com/angular-ui/angular-google-maps/blob/master/src/coffee/directives/api/managers/marker-manager.coffee#L59

and they became like i posted above.
